### ISSUE TITLE
[Bug/Request] Ability to give orders to mission provided escorts.

### ISSUE THREAD

**Issue by Zitchas** [endless-sky/endless-sky#8240]
### Problem Description

I'm always frustrated when the game gives me escorts in a mission, and they don't obey my orders.

In some cases they make it clear that these are independent ships that just happen to be friendly to me and are I'm supposed to care about, but they are clearly independent ships that are just around to be helpful. I'm fine when ships like that don't obey orders, because they're independent people just following me around.

On the other hand, sometimes ships are specifically entrusted to my command, or are specifically described as being either subordinates, people who have designated me the leader, or put me in charge of their security, or other situations where they should be listening to my orders... Then they should be listening for and obeying my instructions.

That's why I list this as bug/request.

I consider it a bug when ships that should be under my command are not.

But this functionality doesn't seem to exist (even the `escort` personality, which I'd honestly expect to include it), so this could also be considered a request.

### Related Issue Links

Discord discussion

### Desired Solution

Probably the best solution would be having a personality (or a value for the escort personality) that makes escort ships actually behave *as escorts.* This would basically involve two specific things: a) they would obey the players commands (gather, hold, attack, etc.) and b) they would be hostile/friendly/etc to things based on the player's reputation.

For optimum flexibility, perhaps have varying levels of obedience? For instance, merchant escort missions should probably have the merchant obey gather and hold commands, but would ignore "attack" commands. Combat ships placed under your command, however, should obey *all* your commands.

### Alternative Approaches

We could just make the escort personality by default obey escort commands. It'd make sense, and is sort of what the player would expect from having the ships listed in their escort area.

Maybe have this as the default behavior, with a specialized "tactical tracker" personality that is used for situations where we want the player to have access to the tactical info on a ship, but without being able to give it orders (basically how "escort" works now, but preferably with some kind of display differentiation to make it clear it's *not* an escort and won't obey orders)

### Additional Context

A *lot* of our fleet battles require us to keep ships alive either in order to assist in other battles, or becaues their survival is an important story requirement. In my current playthrough I've taken to intentionally taking these fleets on round-about routes in order to get them stranded in uninhabited systems as a means of getting them to "hold position" so that I could scout ahead or engage the enemy before they arrived (for instance, to use up the enemy's missiles in a ship that can dodge them.

We should not be having to forcibly strand ships that are nominally following our directions in order to get them to do stuff that we'd normally expect them to do if asked.

**Comment by Hurleveur** [comment]
For the hostility part thats already what happens when you use the escort government; they fight your enemies and act like on of your escorts except that indeed you cant order them around.

I like the idea of expanding the escort personality instead of adding a new one

**Comment by Zitchas** [comment]
I was under the impression it was possible to assign escorts to a government that the player has a negative reputation with, and thus be hostile to them...

**Comment by Hurleveur** [comment]
Yes, but I was speaking of if you give them the "Escort" government. You can shoot them and nothing will happen

**Comment by vitalchip** [comment]
I think 'escorts' need to be sub divided.

1. Escorts which have hired the player.  These don't follow your commands.

This would be the likes of a merchant escort mission, they've hired you so what should happen here is that they head for the destination and it's your job to follow them and keep them safe.

2. Escorts the player hires or when the mission states you are the leader. These follow your commands (to a point)

The player should be able to hire ships to protect them (imagine for a moment that it wasn't incredibly easy to amass your own large fleet)  and these ships should follow every order unless it leads them to death or attack a friendly etc

For missions I would prefer a subset of orders, almost all missions are already too easy to complete apart from where some NPC (Danforth etc) comes along but has to be saved. This is usually the only tricky part because they are as dumb as rocks. Hold shouldn't be an option because all the player will do is keep the NPC in another system, further trivialising the mission. Likewise MoveTo should be disallowed. Gather probably ok as a fleet wide order.  You should be able to group the NPC with a couple of your own escorts (to protect it) and select an easy target for it though.

3. Escorts with a common goal but neither of the above. these generally won't follow commands but they might respect a fleet wide command to gather.

These are just incidental ships, no reason they'd take your orders but if they see everyone else is falling back to regroup they might do the same.

**Comment by 1010todd** [comment]
For extra flexibility perhaps each npc could have their own list of what command they'll follow like:
```
# Hired as escort
npc 
	government "Merchant"
	...
	"allowed commands"
		"focus fire"
		"move to"
		"gather"
	fleet "Bounty Hunter"

# Common goal
npc
	government "Free Worlds"
	...
	"allowed commands"
		"gather"
	fleet "Large Free Worlds"
	
# Devoted follower/Hacked robot or something.
npc
	government "Merchant"
	...
	"allowed commands"
		"focus fire"
		"move to"
		"gather"
		"suicidal"
		"friendly fire"
```
Or if possible a more granular `if`/`to allow` conditions for them to follow each command would be very awesome, like.
```
	"allowed commands"
		"move to"
			to allow
				or
					is idle #not doing anything (ik we don't have this, just some idea)
					"reputation: Free Worlds" >= 1000
					
```
Though these are just some random idea that'll be really cool to see but I can see how much more complicated this will be compared to just allowing npcs to follow some specific commands.

**Comment by Hurleveur** [comment]
The granular conditions could always be added later but definitely shouldn't be done straight away I think, or it would complexify the PR too much, especially since the situation on what we do with the player's conditions is unclear

**Comment by Zitchas** [comment]
For escorts missions/jobs, I see three main varieties. All three of these are perfectly valid reasons for an NPC to hire the player, and should exist in the game, but right now we effectively only have scenario 2 that feels kind of like 3, no matter what the mission text actually happens to say.
1. Player is hired to protect the ship as it goes from A to B.
2. Player is hired to guide the ship from A to B.
3. Player is hired to provide cover/distraction as ship goes from A to B.

Each of these would have different AI behaviors and order acceptance.
1. NPC flies from A to B, stopping for fuel as necessary. Player's nav routing and hyperjump activation is synced to the NPC's and automatically triggered when they jump. They can change/break it, but doing and leaving the NPC alone in a system for more than a specified X amount of time fails the mission. These NPCs don't follow orders at all,  but instead will issue orders to the player. "Gather to me! hails" and "attack the XYZ" messages, etc, which the player can act on as desired.
2. NPC follows the player. They should abide by gather, goto, and hold commands, with the exception that when the player hyperjumps, these will always follow, never to be left behind.
3. Everyone's independent. The NPC flies from A to B on their own, and the player can fly as they wish, with the caveat that the NPC has to be present when the player lands at B in order to complete. Probably obey gather, hold, and goto, but they go back to their regular routefinding once the player leaves the system.


As far as military missions go... The amount of orders they take should match the mission text. If ships are being assigned to the player's command/leadership/etc., then they should follow commands. This is particularly true if ships are being sent along to help the player.

If the content creators deem the player needing help to complete a battle and include a fleet following the player's lead, then that fleet should obey orders.

On the other hand, some military missions should have the player being just members of someone else's fleet, and it acts basically like scenario 1 where their nav system is tied to the NPC, and the NPC will issue "focus fire" instructions whenever they have a target selected, etc. 

As an aside, this alone would probably make the frustrations with NPC fleets *massively* better: If the NPC flagships - at least those being commanded by named tactically-intelligent NPCs - actually used the "Focus fire" order and were obeyed by the rest of the fleet (and told the player, too, so the player can focus fire as well.

Now, if the NPC is along to observe or some other story reason that requires them to stick with the player but not obey them, then they need to have a more intelligent personality set. For instance, an observer ship might flee to the edges of the system.  Or have a general "stay away from enemies" or the like personality based on the character in question.

**Comment by vitalchip** [comment]
My concern around Move and Hold orders is that they'll just be used to send any NPC to the far end of the system, effectively removing them from the equation. we all know we need them to act a bit better but being able to send them away probably isn't the right tack.

I suppose really that when a 'fighting' NPC comes with you it should either just be there to help but also dispensible or it should be something which actually has a meaningful contribution to make.

**Comment by Zitchas** [comment]
The other elephant is that, even if the player is flying solo, the NPCs that are theoretically there to help just end up being a giant RNG lottery: Maybe they'll win the battle for you, maybe they'll lose the battle for you, and in both case the player has virtually zero impact!

That's really the part that bugs me the most, I think. Whether or not I lose the battle shouldn't be boiling down to the RNG of whether the fleet I'm with happens to luck into behaving strategically or not. And that's why on my 3rd or 4th reload, my very first course of action is to figure out how to strand any "must keep alive" escorts somewhere so that they *don't* help me. 95% of the time, getting rid of the "help" is the key to victory.

I know the logic is that if a player is having trouble, if they reload and repeat the battle often enough they'll eventually luck into a win. But that's frustrating, as well as problematic when it is also serving as a block to frequently prevent the player from winning. It's basically ensuring that the player will usually need to reload at that point, instead of  only needing to reload if they die.

For such situations, I'd much rather we implemented some kind of scaling mechanic. As in, a mechanic whereby we track how many times the player has failed a mission, and incorporate ways of tweaking it incrementally.

For example, if we had such a "fail" tracker, we could use conditions inside the mission to change the NPC deployments:
- has 0 fails: 4 allied ships, 4 enemy ships, regular system spawns (assuming a FW battle with both FW and navy spawns)
- has 1 fail: 5 allied ships, 4 enemy ships, regular system spawns.
- has 2 fails: 5 allied ships, 4 enemy ships, regular friendly spawns, but hostile spawn time increased by 25%.
- has 3 fails: 5 allied ships, 3 enemy ships, regular friendly spawns, but hostile spawn time increased by 50%.
- has 4 fails: 6 allied ships, 3 enemy ships, friendly spawn frequency shortened by 25%, hostile spawn time increased by 50%.
- has 5 fails: 6 allied ships, 3 enemy ships, friendly spawn frequency shortened by 25%, no hostile spawns.
- etc. 

Basically, the point being that if the fail count is a condition that missions can access, then we could probably use it without too much trouble to let the content creator adjust the difficulty by small steps depending on the scenario. 


As far as that goes, no matter the player's rank. Even if they're the basic military grunt and everyone is higher rank than them, if they have a reputation for winning solo battles against great odds... Chances are even their superiors are going to be willing to hang back and just let them do their thing. It's not the superiors saying that they can give them orders, it's them acknowledging that the player has a recognized skill at doing solo fights, so let them do it solo.

**Comment by vitalchip** [comment]
For scenario 1 (the Captain of.. escort to..) I've been playing the following way:

When you leave the planet the mission escorts line up to jump to the next system, they'll shoot at anything hostile while doing so but won't move to engage or try to keep station with the player, jumping is their priority.

This leaves the player free to engage hostiles or shield the escort without 'dragging them along'  the player must wait for the escorts to jump before following, if the player jumps first the escorts will NOT follow.

When the escorts arrive in a system they stop and wait for the player, once the player arrives they will either jump to the next system or land (if destination or refuel).

What this means is that the player cannot simply jump ahead, they actually have to protect the escorts in each system along the way. Due to the way I have things set up this might mean covering the ships while they move through each system to the jump departure point, so genuine escorting. 

If the flagship is fast it can wait until the escorts jump, then overtake the escorts in hyperspace to arrive first and 'clear the streets' before the escorts arrive.

I haven't linked navigation back to the player as there often isn't much time to choose which ship you want to sync with in the first system. What I've done is put the escort jump destination into the escorts display, so you know which way they are going.

<img width="306" alt="Screenshot 2023-01-31 at 16 57 26" src="https://user-images.githubusercontent.com/65342577/215837038-3553c75e-2e6a-4aa4-955c-d21867448526.png">

**Comment by Zitchas** [comment]
That sounds like a pretty good system.

**Comment by Hurleveur** [comment]
About the change with failures I dont think its such a good idea. Instead I think the powers of the fleet should depend on the strength of the player, which is why I made an issue about making conditions usable everywhere instead of raw numbers, that UnorderedSigh already has nicely taken on for the fleet part.

**Comment by Zitchas** [comment]
Well, that's the problem: What's an appropriate challenge for me at X fleet strength may be insurmountable to one person and extremely easy to a third person. Having enemy fleet strengths adjusted based on player's strength, set by the content creator, basically ensures that the challenge is always calibrated to whatever the content creator thinks is a good challenge. If the player is worse at combat than the content creator, though... Then they're stuck, since any attempt at making themselves more powerful (adding another ship to their fleet, etc.) will be reflected by a tougher challenge, which is counter-productive.

On the other hand, having a record of failures and adjusting the difficulty of that encounter based on how many times the player has failed it lets the content creator set the challenge... And if the player succeeds, great. If not, then the challenge level is iteratively lowered until the player can pass it. And those iterations could  be extra minor, almost insignificant, but it might be all it takes.

**Comment by Hurleveur** [comment]
Well I dont see why we couldnt have both
But I think that the lower chance on failure could just be replaced by a gamerule difficulty setting that you can adjust as you see fit.

**Comment by Zitchas** [comment]
Sure, to an extent.

1. At the initial approach, have strength calculations setting the battle to the appropriate difficulty according to the author.

2. In the preferences, have a difficulty preference:
- Normal (adjusts depending on strength and number of retries)
- Max (just nails the difficulty to the maximum fleet strength values)
- Min (nails the difficulty to the minimum fleet strength with whatever the max level of failure accommodation is)

3. Perhaps also have a preference for "how fast should the game ease up challenges after failures?" fast, moderate, slow, never. (slow the easing wouldn't kick in until after 6 failures, and each step down would take 3 failures at that level before making it another step easier. Moderate wouldn't kick in until after 4 failures, and it would take 2 failures at each level to push it down another step. Fast would kick in after 2 failures, and would take 1 failure at each level to push it down another step. Never would basically be a "disable difficulty adaptation.")

4. In the battle proper have a failure counter that keeps track of how often people fail a battle. Then establish a bunch of difficulty levels ranging from slight. (enemy reinforcements are slightly slower to spawn) to major (extra friendly ships and fewer enemy ships and no enemy reinforcements). At least 3 difficulty levels, could theoretically be infinite, probably 3-5 would be more than sufficient.

**Comment by Amazinite** [comment]
> Is the request here similar to [#3506](https://github.com/endless-sky/endless-sky/issues/3506)?

Seems like it. Closing this issue in favor of tracking this on the older one.
